src/realm/bptree.hpp

```diff
@@ -438,9 +438,10 @@ inline BpTreeNode& BpTreeBase::root_as_node()
 
 inline const BpTreeNode& BpTreeBase::root_as_node() const
 {
+    Array* arr = m_root.get();
     REALM_ASSERT_DEBUG(!root_is_leaf());
-    REALM_ASSERT_DEBUG(dynamic_cast<const BpTreeNode*>(m_root.get()) != nullptr);
-    return static_cast<const BpTreeNode&>(root());
+    REALM_ASSERT_DEBUG(dynamic_cast<const BpTreeNode*>(arr) != nullptr);
+    return static_cast<const BpTreeNode&>(*arr);
```

Comment:
Just to clarify (I think I wrote the ancestral version of this code): The `dynamic_cast` is meant only as a check in the assertion. `static_cast` is faster when you know it can never fail, and this is ensured by surrounding invariants (which happened to be broken as an effect of this bug). :-)